Frequently requested changes: add bypassing visibility (#323)
rust-lang/lang-team: src/frequently-requested-changes.md

Cross-referencing to other discussions:

* <https://github.com/rust-lang/rfcs/issues/1397>
* <https://github.com/rust-lang/rust/issues/17027>
* <https://github.com/rust-lang/unsafe-code-guidelines/issues/176>

## A way to bypass visibility, including an `unsafe` bypass

> Items are only accessible if they are marked `pub` or re-exported as such;

Reviewer: This line doesn't read correctly to me as stated, i.e. "or re-exported as such" seems too strong.  E.g.:

```rust
mod m {
    fn f() {}
}
pub use m::f; //~ ERROR function `f` is private
```

Author: I see, I was under the impression you could do something like that. I agree it doesn't make sense as written.

they are otherwise private by default. People sometimes wish to break that
rule to access internals of libraries they're using, for example to access
private fields of a type or to call private functions.

> This could break invariants assumed by the crate's author, which if any
> unsafe code depends on those could lead to undefined behavior.

Reviewer 2: ```suggestion
This could break invariants assumed by the crate's author, which, if any
unsafe code depends on those, could lead to undefined behavior.
```

> More importantly, allowing people to violate privacy would destroy SemVer.
> If people can access and use implementation details of other crates then
> that means that almost any change is now a breaking change. This would lead
> to widespread fallout across the crate ecosystem.

Reviewer: ```suggestion
It's also hard to see how allowing privacy violations could be made
compatible with SemVer.
```

> More importantly, allowing people to violate privacy would destroy semver.
> If people can access and use implementation details of other crates then
> that means that almost any change is now a breaking change. This would lead
> to widespread fallout across the crate ecosystem.
> 
> Making it `unsafe` does nothing to prevent these issues. `unsafe` is
> used to deal with memory safety problems and it is not in any way useful to
> deal with semver concerns.

Reviewer 2: Cargo documentation more often (though not entirely consistently) uses “SemVer” instead of “semver”.

```suggestion
More importantly, allowing people to violate privacy would destroy SemVer.
If people can access and use implementation details of other crates then
that means that almost any change is now a breaking change. This would lead
to widespread fallout across the crate ecosystem.

Making it `unsafe` does nothing to prevent these issues. `unsafe` is
used to deal with memory safety problems and it is not in any way useful to
deal with SemVer concerns.
```

> Making it `unsafe` does nothing to prevent these issues. `unsafe` is
> used to deal with memory safety problems and it is not in any way useful to
> deal with SemVer concerns.

Reviewer: ```suggestion
This does not fit with our model of `unsafe` in Rust.
```

Forking a crate (to insert the necessary `pub`s) does not have these

> problems. As such a better way to achieve this would be to make patch

Reviewer 2: ```suggestion
problems. As such, a better way to achieve this would be to make patch
```

> Forking a crate (to insert the necessary `pub`s) does not have these
> problems. As such, a better way to achieve this would be to make patch
> dependencies more ergonomic to use and maintain.

Reviewer: ```suggestion
Forking a crate (to insert the necessary `pub`s) does not have these
problems, and so making patch dependencies more ergonomic to use and
maintain could be one alternative to explore.
```
